apps/test-app/frontend/src/components/tree-widget/stateless-v2/TreeActions.ts

```diff
@@ -0,0 +1,292 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import { Draft, enableMapSet, produce } from "immer";
+import { EMPTY, Observable, reduce, Subject, takeUntil } from "rxjs";
+import { HierarchyNode, HierarchyProvider } from "@itwin/presentation-hierarchy-builder";
+import { HierarchyLoader, IHierarchyLoader, LoadedHierarchyPart } from "./TreeLoader";
+import {
+  addHierarchyPart, expandNode, isTreeModelHierarchyNode, removeSubTree, TreeModel, TreeModelHierarchyNode, TreeModelRootNode,
+} from "./TreeModel";
+import { PresentationHierarchyNode, PresentationTreeNode } from "./Types";
+
+enableMapSet();
+
+/** @internal */
+export interface TreeState {
+  model: TreeModel;
+  rootNodes: Array<PresentationTreeNode> | undefined;
+  isLoading: boolean;
+}
+
+/** @internal */
+export class TreeActions {
+  private _loader: IHierarchyLoader;
+  private _currentModel: TreeModel = { idToNode: {}, parentChildMap: new Map(), rootNode: { id: undefined, nodeData: undefined } };
+  private _disposed = new Subject<void>();
+
+  constructor(private _updater: (actionOrValue: TreeState | ((initialState: TreeState) => TreeState)) => void) {
+    this._loader = new NoopHierarchyLoader();
+  }
+
+  private getModel(model: TreeModel, actionOrModel: ((model: Draft<TreeModel>) => void) | TreeModel) {
```

Comment:
not a fan of this `getModel` name - could we use "produce", "create", "build" instead of "get"?

Reply:
maybe also rename the first arg to "seed"?

Reply:
Renames to `produceModel`.